### PR TITLE
Alex differential drive test

### PR DESCRIPTION
**Pull request by AlexanderTsarapkine**
Added three driving modes (Curvature, Arcade, and tank drive) with ability to change between them through accessing the RobotContainer class. Joystick inverts can be modified in config.

Maps to #42

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -52,7 +51,14 @@ private void configureButtonBindings() {
     driverStick = new Joystick(0);
     controlStick = new Joystick(1);
 
-    driveCommand = new ArcadeDriveWithJoystick(driverStick, Config.LEFT_CONTROL_STICK_Y, Config.INVERT_FORWARD, Config.RIGHT_CONTROL_STICK_X, Config.INVERT_SIDE);
+    /**
+     * Select different driveCommands to control the robot in different modes
+     */
+
+    //driveCommand = new ArcadeDriveWithJoystick(driverStick, Config.LEFT_CONTROL_STICK_Y, Config.INVERT_FIRST_AXIS, Config.RIGHT_CONTROL_STICK_X, Config.INVERT_SECOND_AXIS);
```
Comment (Zinka010) [inline src/main/java/frc/robot/RobotContainer.java]: You should remove the commented out code if it serves no purpose

File: src/main/java/frc/robot/commands/ArcadeDrive.java [hunk 2]
```diff
@@ -22,11 +25,18 @@
...
+ * Creates the arcade drive
+ * 
+ * @param forwardVal The values to use for driving forward
+ * @param rotateVal the values to use for rotating
+ * @param squareInputs Whether or not to square the forward and rotation values
+ * @param initBrake whether or not to start and end the command in brake or coast mode
+ */
   public ArcadeDrive(Supplier<Double> forwardVal, Supplier<Double> rotateVal, boolean squareInputs, boolean initBrake) {
 
+    //Ensure that this command is the only one to run on the drive base
```
Comment (Zinka010) [inline src/main/java/frc/robot/commands/ArcadeDrive.java]: Nitpick: Normally we use a space after `//`

File: src/main/java/frc/robot/commands/ArcadeDrive.java [hunk 3]
```diff
@@ -46,13 +56,15 @@ public void initialize() {
   // Called every time the scheduler runs while the command is scheduled.
   @Override
   public void execute() {
+    //pass values to drive base to make the robot move
```
Comment (Zinka010) [inline src/main/java/frc/robot/commands/ArcadeDrive.java]: Nitpick: Space with a capital letter as well after the `//`

File: src/main/java/frc/robot/commands/CurvatureDrive.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    private final Supplier<Double> forwardVal;
+    private final Supplier<Double> curveSpeed;
+    private final boolean initBrake;
+    private final Supplier<Boolean> buttonPress;
+    private final boolean squareInputs;
+
+    /**
+     * Creates the arcade drive
+     *
+     * @param forwardVal The values to use for driving forward
```
Comment (Zinka010) [inline src/main/java/frc/robot/commands/CurvatureDrive.java:23]: Make the Javadoc represent this constructor, not the one from ArcadeDrive

File: src/main/java/frc/robot/commands/CurvatureDrive.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                DriveBase.getInstance().setBrakeMode(true);
+            }
+
+            DriveBase.getInstance().curvatureDrive(forward * 0.6, (override ? rotation / 2.5 : rotation), override);
+        } else {
+            if (DriveBase.getInstance().isBrakeMode()) {
+                DriveBase.getInstance().setBrakeMode(false);
+            }
+
+            DriveBase.getInstance().curvatureDrive(forward, (override ? rotation / 2 : rotation), override);
```
Comment (Zinka010) [inline src/main/java/frc/robot/commands/CurvatureDrive.java:72]: Great stuff here!

File: src/main/java/frc/robot/commands/TankDriveWithJoystick.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+   * 
+   * @param joy1 The First Joystick
+   * @param axis1 The axis value of the joystick
+   * @param invert1 Weather or not to invert the first axis value
+   * @param joy2 The Second Joystick
+   * @param axis2 the axis value of the second joystick
+   * @param invert2 Weather or not to inver the second axis2 value
+   */
+  public TankDriveWithJoystick(Joystick joy1, int axis1, boolean invert1, Joystick joy2, int axis2, boolean invert2) {
+
```
Comment (Zinka010) [inline src/main/java/frc/robot/commands/TankDriveWithJoystick.java]: Nitpick: remove general unnecessary whitespace

File: src/main/java/frc/robot/sensors/AnalogSelector.java [hunk 7]
```diff
@@ -1,122 +0,0 @@
-package frc.robot.sensors;
```
Comment (Zinka010) [inline src/main/java/frc/robot/sensors/AnalogSelector.java]: Why was this deleted?

**Review by Zinka010 — CHANGES_REQUESTED** [review]
Overall it's good but here are a few things to change/consider?

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (AlexanderTsarapkine) [reply src/main/java/frc/robot/RobotContainer.java]: Wouldn't it be convenient to leave the commented code it since we can just uncomment the mode we want to use and comment out the one we don't so we can do quick changes?

**Review by georgetzavelas — COMMENTED** [review]
You did a lot of great work here!  Looks like you need to rebase your changes.

Something to consider (you don't have to do this just stating it as a learning opportunity): A rule of thumb that I like to use is to try and keep your PRs smaller rather than larger. The more code you merge in one PR, the more chance for a bug to sneak in and your reviewer to not see it since they have a lot of code to review. For example this PR could be broken into 3 PRs:
- CurvatureDrive + CurvatureDriveWithJoystick
- TankDrive + TankDriveWithJoystick
- The rest of your changes that then use the new classes

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (kevlam2706) [reply src/main/java/frc/robot/RobotContainer.java]: Convenient maybe but kinda messy. Part of what we're doing here is trying to emulate real software engineering best practices. Once we know what the drive team wants, this command won't really change, so need to keep other versions on "standby".

**Review by kevlam2706 — APPROVED** [review]
Why do my diffs show that some files were deleted?

Looks good based on other review comments.

Looks like you have some merge conflicts as well.

Please resolve ASAP and get your stuff into master so others can start using it.

Did you test everything thoroughly?

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (AlexanderTsarapkine) [reply src/main/java/frc/robot/RobotContainer.java]: Alright

File: src/main/java/frc/robot/RobotContainer.java [hunk 8]
```diff
@@ -57,14 +52,12 @@ private void configureButtonBindings() {
...
-    driveCommand = new ArcadeDriveWithJoystick(driverStick, Config.LEFT_CONTROL_STICK_Y, Config.INVERT_FORWARD, Config.RIGHT_CONTROL_STICK_X, Config.INVERT_SIDE);
+    /**
+     * Select drive mode for robot
+     */
 
+    driveCommand = new ArcadeDriveWithJoystick(driverStick, Config.LEFT_CONTROL_STICK_Y, Config.INVERT_FIRST_AXIS, Config.RIGHT_CONTROL_STICK_X, Config.INVERT_SECOND_AXIS);
     DriveBase.getInstance().setDefaultCommand(driveCommand);
-
-    // Instantiate the intake command and bind it
-    intakeCommand = new OperatorIntakeCommand();
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/RobotContainer.java:66]: Is there a reason these 2 lines are being deleted?

**Review by Zinka010 — APPROVED** [review]
Good stuff

File: src/main/java/frc/robot/commands/CurvatureDriveWithJoystick.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+     * Drives using axes that are on separate joysticks
+     *
+     * @param joy1    The joystick that the first axis is on
+     * @param axis1   The first axis
+     * @param invert1 Whether to negate the first joystick input
+     * @param joy2    The joystick that the second axis is on
+     * @param axis2   The second axis
+     * @param invert2 Whether to negate the second joystick input
+     */
+    public CurvatureDriveWithJoystick(Joystick joy1, int axis1, boolean invert1,
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/commands/CurvatureDriveWithJoystick.java:20]: If the values for invert are always the same you should consider not passing them and simply inferring them here.

File: src/main/java/frc/robot/commands/TankDrive.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    //Pass values to drive base to make the robot move
+    DriveBase.getInstance().tankDrive(leftSpeed.get(),rightSpeed.get(),squareInputs);
+  }
+
+
+  @Override
+  public void end(boolean interrupted) {
+    //go back to disabled mode
+    DriveBase.getInstance().setDisabledMode();
+
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/commands/TankDrive.java:59]: 2 returns here?

File: src/main/java/frc/robot/commands/TankDrive.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  protected TankDrive(Supplier<Double> leftSpeed, Supplier<Double> rightSpeed, boolean squareInputs, boolean initBrake) {
+    //Ensure that this command is the only one to run on the drive base
+    //Requires must be included to use this command as a default command for the drive base
+    addRequirements(DriveBase.getInstance());
+
+    this.leftSpeed = leftSpeed;
+    this.rightSpeed = rightSpeed;
+    this.squareInputs = squareInputs;
+    this.initBrake = initBrake;
+
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/commands/TankDrive.java:36]: return?

File: src/main/java/frc/robot/commands/CurvatureDrive.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+        double rotation = Math.abs(curve) < Config.CONTROLLER_DEADBAND ? 0 : curve;
+        boolean override = Math.abs(forward) < Config.CURVATURE_OVERRIDE;
+
+        if (buttonPress.get()) {
+            if (!DriveBase.getInstance().isBrakeMode()) {
+                DriveBase.getInstance().setBrakeMode(true);
+            }
+
+            DriveBase.getInstance().curvatureDrive(forward * 0.6, (override ? rotation / 2.5 : rotation), override);
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/commands/CurvatureDrive.java:66]: 0.6 and 2.5 are magic numbers, can we make these into constants so that the name of the constant will tell us what they do?

**Review by georgetzavelas — APPROVED** [review]
Great work! I love the javadoc usage.

File: src/main/java/frc/robot/commands/TankDrive.java (hunk 10 above)
Comment (AlexanderTsarapkine) [reply src/main/java/frc/robot/commands/TankDrive.java]: This is a void method could you elaborate maybe?

Comment (AlexanderTsarapkine) [reply src/main/java/frc/robot/commands/TankDrive.java:59]: Could you elaborate?

File: src/main/java/frc/robot/RobotContainer.java (hunk 8 above)
Comment (AlexanderTsarapkine) [reply src/main/java/frc/robot/RobotContainer.java:66]: No I think that was an accident I will re add
